zio-http/src/main/scala/zio/http/endpoint/internal/EndpointServer.scala

```diff
@@ -28,10 +29,19 @@ private[endpoint] final case class EndpointServer[R, E, I, O, M <: EndpointMiddl
   private val handler  = single.handler
 
   def handle(request: Request)(implicit trace: Trace): ZIO[R, Nothing, Response] = {
-    endpoint.input.decodeRequest(request).orDie.flatMap { value =>
-      handler(value).map(endpoint.output.encodeResponse(_)).catchAll { error =>
-        ZIO.succeed(single.endpoint.error.encodeResponse(error))
+    endpoint.input
+      .decodeRequest(request)
+      .flatMap { value =>
+        handler(value).map(endpoint.output.encodeResponse(_)).catchAll { error =>
+          ZIO.succeed(single.endpoint.error.encodeResponse(error))
+        }
       }
-    }
+      .catchSomeCause {
+        case Cause.Fail(underlying: HttpCodecError, _) =>
```

Comment:
This breaks the alternatives handling for routes. See `Routes#dispatch`. I think `dispatch` is the place where we should turn this into a response. Dispatch currently returns a `ZIO[R, Nothing, Response]` but I think we could turn this into `ZIO[R, Response, Response]` and have a `BadRequest` response in the error channel of that zio, in case the `HttpCodecError.isHttpCodecError(cause)` is true.


wdyt @jdegoes @vigoo ?

Reply:
Can't `handle` here return `ZIO[R, Response, Response]`, before it ends up as a route?

Reply:
Not 100% sure what you mean with `before it ends up as a route?`. But I guess that happens already before?
If handle would return `ZIO[R, Response, Response]` I think it would also break `dispatch`, at least in the current form. If we would assume, that an Response in the error channel always means, we had an `HttpCodecError` and we can try the next alternative, then I guess we could change the impl. in dispatch, but it might be fragile to future changes, that actually want to return this error response instead of trying an alternative.
I actually think the best argument for moving this to dispatch is, that only after all alternatives are tried your need a error response. So creating responses here, would end up garbage, if there is an alternative. But leaving this as a zio failure with a cause, makes it possible to add up all causes and give an error response of all the `HttpCodecError`s that happened, instead of just the last one.
And creating only responses that would actually be sent out to the caller, feels also more right.

Reply:
Yep, it makes sense, thanks Nabil, I'll move it up to dispatch

Reply:
In this case should list all the alternatives I tries in the response body? Or should just return a simple 400 with no body?
@jdegoes

Reply:
I'd say added the errors should be fine. 1. You want to know the error. 2. There shouldn't be any information leaked except for what was in the request you sent anyway.

Reply:
Actually it would somewhat leak all the possible ways it failed so that's why I didn't add it